doc/specl.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.6.
.TH SPECL "1" "February 2023" "specl (Specl) 14.1.8" "User Commands"
.SH NAME
specl \- Specl
.SH SYNOPSIS
.B specl
[\fI\,OPTION\/\fR]... [\fI\,FILE\/\fR]...
.SH DESCRIPTION
Behaviour Driven Development for Lua.
.PP
Develop and run BDD specs written in Lua for RSpec style workflow, by verifying
specification expectations read from given FILEs or standard input, and
reporting the results on standard output.
.PP
If no FILE is listed, then run specifications for all files from the 'specs/'
directory with names ending in '.yaml'.
.PP
Where '\-' is given as a FILE, then read from standard input.
.TP
\fB\-\-help\fR
print this help, then exit
.TP
\fB\-\-version\fR
print version number, then exit
.TP
\fB\-\-color\fR=\fI\,WHEN\/\fR
request colorized formatter output [default=yes]
.TP
\fB\-\-coverage\fR
generate coverage report, if luacov is installed
.TP
\fB\-1\fR, \fB\-\-fail\-fast\fR
exit immediately on first failed example
.TP
\fB\-f\fR, \fB\-\-formatter\fR=\fI\,FILE\/\fR
use a specific formatter [default=progress]
.TP
\fB\-\-unicode\fR
allow unicode in spec files
.TP
\fB\-v\fR, \fB\-\-verbose\fR
request verbose formatter output
.PP
Filtering:
.TP
\fB\-e\fR, \fB\-\-example\fR=\fI\,PATTERN\/\fR
check only the examples matching PATTERN
.TP
+NN
check only the example at line NN in next FILE
.TP
FILE:NN[:MM]
check only the example at line NN in this FILE
.PP
When filtering by PATTERN, an example is considered a match if PATTERN matches
the concatenation of nested YAML descriptions leading directly to that example
in its spec file.
.PP
You can specify +NN (where NN is a line number) multiple times for the next
specified FILE, or interspersed to specify different line filters for different
FILEs. Specifying any number of +NN will prevent all specifications in that
file except those selected by a +NN filter from being checked. If +NN is not
the first line of an example (as would be displayed by a failing example in
verbose mode), the option selects no examples.
.PP
The alternative FILE:NN:MM syntax makes it easy to cut and paste from Specl
failure output, but allows only a single line NN to be filtered (except when
combined with +NN filters).  The optional :MM suffix is ignored \fB\-\-\fR and merely
represents the ordinal number of an `expect` statement in a particular example
in verbose Specl formatter outputs.
.PP
Due to a shortcoming in libYAML, unicode characters in any passed FILE prevent
Specl from working. The '\-\-unicode' option works around that shortcoming, but
any error messages caused by Lua code in FILE will usually report line\-numbers
incorrectly.  By default, errors report accurate line\-numbers, and unicode
characters are not supported.
.SH AUTHOR
Written by Gary V. Vaughan <gary@gnu.org>, 2013
.SH "REPORTING BUGS"
Report bugs to http://github.com/gvvaughan/specl/issues.
.SH COPYRIGHT
Copyright \(co 2023, Gary V. Vaughan
Specl comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of Specl under the terms of the GNU
General Public License; either version 3, or any later version.
For more information, see <http://www.gnu.org/licenses>.


NEWS
# Specl NEWS - User visible changes.

## Noteworthy changes in release 14.1.3 (2015-10-26) [stable]

### Bug fixes:

  - internal `deepcopy` no longer overflows the stack when copying
    tables that are their own metatable across nested Specl
    environments.


## Noteworthy changes in release 14.1.2 (2015-08-08) [stable]

### Bug fixes:

  - `equal` matcher really does compare mutable keys correctly,
    which works on LuaJIT and Lua 5.3 too.


## Noteworthy changes in release 14.1.1 (2015-08-02) [stable]

### Bug fixes:

  - `equal` matcher compares mutable table keys correctly, so things
    like this now work:

    ```lua
    expect ({[{k="mutable"}]=true}).to_equal ({[{k="mutable"}]=true})
    ```


## Noteworthy changes in release 14.1.0 (2015-01-03) [stable]

### New features

  - `badargs.diagnose` also accepts '?' in lieu of a proper function
    name in argument errors (as often returned by LuaJIT).

  - `badargs.format` will produce a `bad field` diagnostic when called
    as:

    ```lua
    badargs.format ("function_name", narg, nil, "field_name")
    ```

  - New matcher `raise_matching` compares an error against a Lua pattern.


## Noteworthy changes in release 14.0.0 (2014-12-31) [stable]

### New features

  - Preliminary Lua 5.3 support.

### Incompatible changes:

  - `badargs.format` now formats message with the word "arguments" for
    all ordinal values except exactly 1, where "argument" is used.
    Previously, Specl erroneously used "argument" for ordinal values of
    less than or equal to 1.

  - Ancient code to replace underscores in example description with
    spaces has finally removed.

### Bug fixes:

  - `specl.inprocess` now propagates function environments correctly,
    so examples inside the `inprocess.call` are not tallied to the
    main specl counters; `formatter.report` does not lose track of what
    descriptions have been displayed after an `inprocess.call`; etc.

  - When expecting automatic execution of specfiles in ./specs/*,
    without luaposix installed, error is diagnosed properly.


## Noteworthy changes in release 13 (2014-10-04) [stable]

### New features:

  - Specl's own modules, and modules loaded from spec-file directories
    are now loaded inside the function environment of the running
    example, and so have access to Specl's Lua extensions, such as
    `expect` and matchers.

  - `inprocess.capture` returns a `Process` object, suitable for
    matching using the `specl.shell` matchers.

  - New `to_raise` matcher as an alias for `to_error` to avoid flagging
    capitalised error messages in error matcher argument with Slingshot
    sanity checks.

  - New programmatic `examples` function for building example groups and
    helper functions directly in Lua, whenever YAML starts to get in the
    way.  See documentation for examples.

  - New `specl.badargs` module provides `format` to generate standard
    argument error messages that can be passed to matchers to ensure
    correct error message formatting:

    ```lua
    expect (posix.write ()).to_raise (format ("write", 1, "int"))
    ```

    And `diagnose`, which writes a full set of expectations to check all
    missing, extraneous and badly typed arguments are diagnosed
    correctly:

    ```yaml
    - context with bad arguments:
        badargs.diagnose (posix.write "(int, string)")
    ```

### Incompatible changes:

  - `inprocess.capture` returns a `Process`, not a pair of strings
    and the call results. Call results are still available in the array
    part of the returned `Process`.

  - Specl now requires latest stdlib v40 to be installed on the module
    path, rather than carrying its own copies of parts of stdlib it wants
    to use.

  - Specl now installs as a selection of library files (in the module
    subdirectory `specl`), and a thin wrapper script that sets up the
    execution environment and then calls `specl.main (arg):execute ()`.
    This simplifies debugging, and remove all the infrastructure for
    expanding `from` macros, and concatenating the resulting sources, as
    well as allowing piece-meal loading of specl modules.

### Bug fixes:

  - Enhanced resolution timers now also work with luaposix > 31.

  - The report formatter would not display subsequent repeats of a
    context description after the first under some circumstances.  It
    now displays context descriptions correctly in that case too.

  - Don't prevent examples from loading a different luaposix than the
    installed default loaded by Specl.  This allows the luaposix specs
    to work in the build tree without priming LUA_CPATH and LUA_PATH
    before calling specl, for example.

  - `inprocess.capture` now supports capturing results of a functable
    call.

  - `specl.shell` matchers now correctly succeed when using `not_`
    prefixes to specify that a `Process` should not contain some error
    or standard output text, and that `Process` stream contains `nil`
    to indicate the stream was not written at all. For example:

    ```lua
    expect (capture (function () print "stdout" end)).
      not_to_contain_error "stderr"
    ```

  - After loading system modules it depends on, Specl removes them
    from the Lua module cache so that other versions of those same
    modules (yaml, posix, posix.sys) can use Specl to check their own
    behaviours by loading the to-be-checked versions from
    `spec_helper.lua`.


## Noteworthy changes in release 12 (2014-04-25) [stable]

### New features:

  - Now that this API is stable, there is full documentation for
    `specl.shell` in doc/specl.md.

  - New `specl.inprocess.capture` function for running examples that
    interact with io streams.  `inprocess.capture` wraps a function call
    so that standard input calls read from the supplied buffer, and any
    output is captured rather than leaked:

    ```lua
    inprocess.capture (function (...)
      print (...)
      io.stderr:write (io.stdin:read "*a")

lib/specl.lua
-- Specification testing framework.
-- Written by Gary V. Vaughan, 2013
--
-- Copyright (c) 2013-2014 Gary V. Vaughan
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 3, or (at your option)
-- any later version.
--
-- This program is distributed in the hope that it will be useful, but
-- WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; see the file COPYING.  If not, write to the
-- Free Software Foundation, Fifth Floor, 51 Franklin Street, Boston,
-- MA 02111-1301, USA.


-- Use the simple progress formatter by default.  Can be changed by run().
local formatter  = require "specl.formatter.progress"
local matchers   = require "specl.matchers"
local std        = require "specl.std"
local util       = require "specl.util"


from std.string import slurp, split


--[[ ================================= ]]--
--[[ Compatibility between 5.1 and 5.2 ]]--
--[[ ================================= ]]--


-- From http://lua-users.org/lists/lua-l/2010-06/msg00313.html
setfenv = setfenv or function(f, t)
  local name
  local up = 0
  repeat
    up = up + 1
    name = debug.getupvalue (f, up)
  until name == '_ENV' or name == nil
  if name then
    debug.upvaluejoin (f, up, function () return name end, 1)
    debug.setupvalue (f, up, t)
  end
  return f
end


loadstring = loadstring or function (chunk, chunkname)
  return load (chunk, chunkname)
end



--[[ ============= ]]--
--[[ Spec' Runner. ]]--
--[[ ============= ]]--


-- Append non-nil ARG to HOLDER.accumulated.
-- If ARG is a table, values for all keys in ARG are accumulated in
-- equivalent HOLDER keys.
-- Used to collect output from formatter calls, to be saved for footer.
local function accumulator (holder, arg)
  if arg ~= nil then
    if type (arg) == "table" then
      holder.accumulated = holder.accumulated or {}
      for k, v in pairs (arg) do
        holder.accumulated[k] = (holder.accumulated[k] or "") .. tostring(v)
      end
    else
      holder.accumulated = (holder.accumulated or "") .. arg
    end
  end
end


-- Access to core functions that we override to run within nested
-- function environments.
local core = {
  load       = load,
  loadfile   = loadfile,
  loadstring = loadstring,
  require    = require,
}


-- A map of module name to global symbol side effects.
-- We keep track of these so that they can be injected into an
-- execution environment that requires a module.
local sidefx = {}


-- Intercept functions that normally execute in the global environment,
-- and run them in the example block environment to capture side-effects
-- correctly.
local function initenv (env)
  for _, intercept in pairs { "load", "loadfile", "loadstring" } do
    env[intercept] = function (...)
      local fn = core[intercept] (...)
      return function ()
        setfenv (fn, env)
        return fn ()
      end
    end
  end

  -- For a not-yet-{pre,}loaded module, try to find it on `package.path`
  -- and load it with `loadstring`, so that any symbols that leak out
  -- (the side effects) are cached, and then copied into the example
  -- block environment, for this and subsequent examples that load it.
  env.require = function (m)
    local errmsg, import, loaded, loadfn

    -- We can have a spec_helper.lua in each spec directory, so don't
    -- cache the side effects of a random one!
    if m ~= "spec_helper" then
      loaded, loadfn = package.loaded[m], package.preload[m]
      import = sidefx[m]
    end

    if import == nil then
      -- Not preloaded, so search package.path.
      if loadfn == nil then
        std.package.mappath (package.path, function (path)
          local filename = path:gsub ("%?", (m:gsub ("%.", "/")))
          local s = slurp (filename)
          if s ~= nil then
            loadfn, errmsg = loadstring (s, filename)
	    return "break"
          end
        end)
      end
      if errmsg ~= nil then return error (errmsg) end

      -- Capture side effects.
      if loadfn ~= nil then
        import = setmetatable ({}, {__index = env})
        setfenv (loadfn, import)
        loaded = loadfn ()
      end
    end

    -- Import side effects into example block environment.
    for name, value in pairs (import or {}) do
      env[name] = value
    end

    sidefx[m] = import
    package.loaded[m] = package.loaded[m] or loaded or nil

    -- Use the system require if loadstring failed (e.g. C module).
    return package.loaded[m] or core.require (m)
  end
end


local run_examples, run_contexts, run


-- Run each of EXAMPLES under ENV in order.
function run_examples (examples, descriptions, env)
  local block = function (example, blockenv)
    local metatable = { __index = blockenv }
    local fenv = { expect = matchers.expect, pending = matchers.pending }

    setmetatable (fenv, metatable)
    initenv (fenv)

    if examples.before ~= nil then
      setfenv (examples.before, fenv)
      examples.before ()
    end

    -- There is only one, otherwise we can't maintain example order.
    local description, definition = next (example)

    if type (definition) == "table" then
      -- A nested context, revert back to run_contexts.
      run_contexts (example, descriptions, fenv)

    elseif type (definition) == "function" then
      -- An example, execute it in a clean new sub-environment.
      table.insert (descriptions, description)

      matchers.init ()

      setfenv (definition, fenv)
      definition ()
      accumulator (formatter,
                   formatter.expectations (matchers.status (), descriptions))
      table.remove (descriptions)
    end

    if examples.after ~= nil then
      setfenv (examples.after, fenv)
      examples.after ()
    end
  end

  for _, example in ipairs (examples) do
    -- Also, run every block in a sub-environment, so that before() and
    -- after() calls from one block don't affect any other.
    local fenv = setmetatable ({}, { __index = env })
    setfenv (block, fenv)
    block (example, fenv)
  end
end


-- Run each of CONTEXTS under ENV in order.
function run_contexts (contexts, descriptions, env)
  for description, examples in pairs (contexts) do
    table.insert (descriptions, description)
    accumulator (formatter, formatter.spec (descriptions))
    run_examples (examples, descriptions, env)
    table.remove (descriptions)
  end
end


-- Run SPECS, according to OPTS and ENV.
function run (specs, env)
  formatter = opts.formatter or formatter

  -- Run compiled specs, in order.
  accumulator (formatter, formatter.header (matchers.stats))
  for _, spec in ipairs (specs) do
    run_examples (spec, {}, env)
  end
  formatter.footer (matchers.stats, formatter.accumulated)
  return matchers.stats.fail ~= 0 and 1 or 0
end



--[[ ----------------- ]]--
--[[ Public Interface. ]]--
--[[ ----------------- ]]--

local M = {
  run       = run,
}


return M
